api/src/main/java/io/druid/segment/loading/DataSegmentPusher.java

```diff
@@ -44,36 +45,48 @@
-   *                        load segments with the same identifier but different contents which is a bad situation. On
-   *                        the other hand, indexers that maintain exactly-once semantics by storing checkpoint data can
-   *                        lose or repeat data if it fails to write a segment because it already exists and overwriting
-   *                        is not permitted. This situation can occur if a task fails after pushing to deep storage but
-   *                        before writing to the metadata storage, see: https://github.com/druid-io/druid/issues/5161.
+   * @param useUniquePath if true, pushes to a unique file path. This prevents situations where task failures or replica
+   *                      tasks can either overwrite or fail to overwrite existing segments leading to the possibility
+   *                      of different versions of the same segment ID containing different data. As an example, a Kafka
+   *                      indexing task starting at offset A and ending at offset B may push a segment to deep storage
+   *                      and then fail before writing the loadSpec to the metadata table, resulting in a replacement
+   *                      task being spawned. This replacement will also start at offset A but will read to offset C and
+   *                      will then push a segment to deep storage and write the loadSpec metadata. Without unique file
+   *                      paths, this can only work correctly if new segments overwrite existing segments. Suppose that
+   *                      at this point the task then fails so that the supervisor retries again from offset A. This 3rd
+   *                      attempt will overwrite the segments in deep storage before failing to write the loadSpec
+   *                      metadata, resulting in inconsistencies in the segment data now in deep storage and copies of
+   *                      the segment already loaded by historicals.
    *
-   *                        If replaceExisting is true, existing objects MUST be overwritten, since failure to do so
-   *                        will break exactly-once semantics. If replaceExisting is false, existing objects SHOULD be
-   *                        prioritized but it is acceptable if they are overwritten (deep storages may be eventually
-   *                        consistent or otherwise unable to support transactional writes).
+   *                      If unique paths are used, caller is responsible for cleaning up segments that were pushed but
+   *                      were not written to the metadata table (for example when using replica tasks).
    * @return segment descriptor
    * @throws IOException
    */
-  DataSegment push(File file, DataSegment segment, boolean replaceExisting) throws IOException;
+  DataSegment push(File file, DataSegment segment, boolean useUniquePath) throws IOException;
```

Comment:
What happens if a deep storage pusher impl doesn't respect `useUniquePath`? This will likely happen for extensions whose authors don't notice the signature change here. It's probably acceptable if what happens is you just get the old behavior.

Reply:
Well, this is a little janky, but note that the signature didn't *really* change, but the meaning of that last boolean parameter did. At first I wanted to force a signature change so that implementors would have to acknowledge the change, but as discussed, decided to make it 'backward-compatible' for the point release. So implementors who don't notice the signature change will get 'replaceExisting=true' behavior for Kafka indexing tasks and 'replaceExisting=false' for all other task types, which seemed reasonable to me (since `replaceExisting=true` was added primarily for the Kafka indexing task type as well)

Reply:
That sounds fair.